### PR TITLE
first_impl

### PR DESCRIPTION
**Pull request by YaTut1901**

### CODE REVIEW

File: src/main/java/main/movie/dao/TicketDao.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package main.movie.dao;
+
+import main.movie.model.Ticket;
+
+public interface TicketDao {
+    Ticket create(Ticket ticket);
```
Comment (Martynice) [inline src/main/java/main/movie/dao/TicketDao.java]: This method should be called "add", but it doesn't matter

File: src/main/java/main/movie/Main.java [hunk 2]
```diff
@@ -58,5 +62,28 @@ public static void main(String[] args) {
...
                 .findAvailableSessions(2L, LocalDate.of(2020,
                         10, 22)));
+
+        UserService userService
+                = (UserService) injector.getInstance(UserService.class);
+
+        User user = new User();
+        user.setPassword("1234");
+        user.setEmail("myemdfail");
+        userService.add(user);
```
Comment (KseniiaMakarova) [inline src/main/java/main/movie/Main.java]: let's register a new user via AuthenticationService
(and add to `register()` method registering the new shopping cart as well)

File: src/main/java/main/movie/Main.java [hunk 3]
```diff
@@ -58,5 +62,28 @@ public static void main(String[] args) {
...
+        System.out.println(userService.findByEmail("myemdfail"));
+        User use2r = new User();
```
Comment (KseniiaMakarova) [inline src/main/java/main/movie/Main.java]: check this variable's name

File: src/main/java/main/movie/dao/TicketDao.java (hunk 1 above)
Comment (KseniiaMakarova) [inline src/main/java/main/movie/dao/TicketDao.java]: I would say that it is nice to be consistent in method names
let's rename this to `add()`

File: src/main/java/main/movie/dao/impl/ShoppingCartDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        try {
+            session = HibernateUtil.getSessionFactory().openSession();
+            transaction = session.beginTransaction();
+            session.update(shoppingCart);
+            transaction.commit();
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("Can't add shoppingCart to DB! " + shoppingCart, e);
```
Comment (KseniiaMakarova) [inline src/main/java/main/movie/dao/impl/ShoppingCartDaoImpl.java]:
```suggestion
            throw new DataProcessingException("Can't update shoppingCart! " + shoppingCart, e);
```

File: src/main/java/main/movie/model/ShoppingCart.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+import javax.persistence.Id;
+import javax.persistence.OneToMany;
+import javax.persistence.OneToOne;
+import javax.persistence.Table;
+
+@Entity
+@Table(name = "carts")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
Comment (KseniiaMakarova) [inline src/main/java/main/movie/model/ShoppingCart.java]: we don't need to generate the id for shopping cart - it should be the same as for user
use `@MapsId` to have only one `id` column in `shopping_cart` table

File: src/main/java/main/movie/model/ShoppingCart.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    private Long id;
+    @OneToMany
+    private List<Ticket> tickets;
+    @OneToOne
+    private User user;
+
+    public ShoppingCart() {
+    }
```
Comment (KseniiaMakarova) [inline src/main/java/main/movie/model/ShoppingCart.java]: do we need this?

File: src/main/java/main/movie/model/Ticket.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
+    private MovieSession movieSession;
+    @ManyToOne
+    private User user;
+
+    public Ticket() {
+    }
```
Comment (KseniiaMakarova) [inline src/main/java/main/movie/model/Ticket.java]: same

File: src/main/java/main/movie/service/impl/ShoppingCartServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    @Override
+    public ShoppingCart getByUser(User user) {
+        return shoppingCartDao.getByUser(user);
+    }
+
+    @Override
+    public void registerNewShoppingCart(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        shoppingCart.setTickets(new ArrayList<>(Collections.emptyList()));
```
Comment (KseniiaMakarova) [inline src/main/java/main/movie/service/impl/ShoppingCartServiceImpl.java]: you can omit this - try to debug and see if null is returned when you look at `tickets` list after you call `ShoppingCart shoppingCart = getByUser(user);` in `addSession()`

also, Collections.emptyList() returns immutable list that can not be changed

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
looks good - let's fix travis so that it would work
check the name of your `.travis.yml` - it should have a dot in the beginning

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**